src/agent/src/rpc.rs

```diff
@@ -997,6 +1006,92 @@ impl protocols::agent_ttrpc::AgentService for AgentService {
         })
     }
 
+    async fn get_ip_tables(
+        &self,
+        ctx: &TtrpcContext,
+        req: GetIPTablesRequest,
+    ) -> ttrpc::Result<GetIPTablesResponse> {
+        trace_rpc_call!(ctx, "get_iptables", req);
+        is_allowed!(req);
+
+        info!(sl!(), "get_ip_tables: request received");
+
+        let mut cmd = "/sbin/iptables-save";
+        if req.is_ipv6 {
+            cmd = "/sbin/ip6tables-save";
+        }
+
+        match Command::new(cmd).output() {
+            Ok(output) => Ok(GetIPTablesResponse {
+                data: output.stdout,
+                ..Default::default()
+            }),
+            Err(e) => {
+                warn!(sl!(), "failed to run {}: {:?}", cmd, e.kind());
+                return Err(ttrpc_error!(ttrpc::Code::INTERNAL, e));
+            }
+        }
+    }
+
+    async fn set_ip_tables(
+        &self,
+        ctx: &TtrpcContext,
+        req: SetIPTablesRequest,
+    ) -> ttrpc::Result<SetIPTablesResponse> {
+        trace_rpc_call!(ctx, "set_iptables", req);
+        is_allowed!(req);
+
+        info!(sl!(), "set_ip_tables request received");
+
+        // create a temp file and write provided ip tables data to the file. If no
+        // data is provided, we expect that the resulting file will be empty, and
+        // the resulting iptables-restore will clear existing rules.
+        let mut file = NamedTempFile::new().unwrap();
+        let _ = match file.write(&*req.data) {
+            Ok(size) => size,
+            Err(e) => {
+                warn!(sl!(), "failed to write iptables to temp file");
+                return Err(ttrpc_error!(ttrpc::Code::INTERNAL, e));
+            }
+        };
+
+        let cmd = if req.is_ipv6 {
+            "/sbin/ip6tables-restore";
+        } else {
+            "/sbin/iptables-restore";
+        };
+
+        let output = match Command::new(cmd)
```

Comment:
You may want to first run the command with `-t`. i.e. test to ensure that it can apply. And if you get a failure return an error. That ensure that you catch any incompatibilities between host and guest versions of iptables.

It also ensure the guest remains consistent.

That also addresses the issue of ensuring that host and guest side iptables binaries are compatible.

Reply:
From my testing, if iptables-restore will fail, the existing iptables are not modified.  In other words, I'll just return an error response over ttrpc if the -t failed, or if the actual attempt failed. End result should be the same?

Reply:
ack. This was just a paranoid check which is lockless